core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/ClusterBootstrapper.java

```diff
@@ -361,4 +373,21 @@ private void validatePollingResultForCancellation(PollingResult pollingResult, S
         }
     }
 
+    String getNodeNameForInstance(InstanceMetaData im, Stack stack,
+            Map<String, AtomicLong> hostGroupNodeIndex, Set<String> clusterNodeNames) {
+        if (isNotBlank(im.getShortHostname())) {
+            return im.getShortHostname();
+        }
+        AtomicLong hostGroupIndex = hostGroupNodeIndex.computeIfAbsent(im.getInstanceGroup().getGroupName(),
+                instantGroup -> new AtomicLong(0));
+        String generatedHostName;
+        do {
+            generatedHostName = hostDiscoveryService.calculateHostname(stack.getCustomHostname(), im.getShortHostname(),
+                    im.getInstanceGroupName(), hostGroupIndex.getAndIncrement(), stack.isHostgroupNameAsHostname());
+        } while (clusterNodeNames.contains(generatedHostName));
```

Comment:
this doesn't scale well at first blink, if I have a 300 node cluster in a single hostgroup and upscale with 50 node what would happen?

Reply:
if 300 node cluster in hostgroup, hostGroupNodeIndex would be initialized to 300 before hitting this loop. So the first new node would get 301 index and subsequent nodes would get next sequential indexes. The existing clusterNodeNames check is failSafe to ensure no duplicate name is generated as that would fail the cluster launch\scaling.